packages/apps/staking/src/components/modals/StakeModal.tsx

```diff
@@ -1,83 +1,178 @@
+import { FC, ChangeEvent, useState } from 'react';
 import {
+  Box,
   Button,
-  CircularProgress,
   InputAdornment,
   TextField,
   Typography,
 } from '@mui/material';
-import React, { useState } from 'react';
-import { useStakeContext } from '../../contexts/stake';
+import { parseUnits } from 'ethers';
+
 import BaseModal from './BaseModal';
+import { ModalError, ModalLoading, ModalSuccess } from '../ModalState';
+import { useStakeContext } from '../../contexts/stake';
+import {
+  useModalRequestStatus,
+  ModalRequestStatus,
+} from '../../hooks/useModalRequestStatus';
 
 type Props = {
   open: boolean;
   onClose: () => void;
 };
 
-const StakeModal: React.FC<Props> = ({ open, onClose }) => {
-  const { handleStake, tokenBalance } = useStakeContext();
+const StakeModal: FC<Props> = ({ open, onClose }) => {
   const [amount, setAmount] = useState('');
-  const [loading, setLoading] = useState(false);
-
-  const handleStakeAction = async () => {
-    setLoading(true);
-    try {
-      await handleStake(amount);
-      onClose();
-      setAmount('');
-    } catch (error) {
-      console.error('Error during staking:', error);
-    } finally {
-      setLoading(false);
+  const [amountError, setAmountError] = useState('');
+  const { handleStake, tokenBalance } = useStakeContext();
+  const { changeStatus, isIdle, isLoading, isSuccess, isError } =
+    useModalRequestStatus();
+
+  const isStakeDisabled = !!amountError || !amount || isLoading;
+
+  const handleClose = () => {
+    setAmount('');
+    setAmountError('');
+    changeStatus(ModalRequestStatus.Idle);
+    onClose();
+  };
+
+  const handleModalAction = async () => {
+    if (isError) {
+      changeStatus(ModalRequestStatus.Idle);
+      return;
+    } else if (isSuccess) {
+      handleClose();
+      return;
+    } else {
+      if (isStakeDisabled) return;
+
+      changeStatus(ModalRequestStatus.Loading);
+      try {
+        await handleStake(amount);
+        changeStatus(ModalRequestStatus.Success);
+        return;
+      } catch (error) {
+        console.error('Error during staking:', error);
+        changeStatus(ModalRequestStatus.Error);
+        return;
+      }
     }
   };
 
-  const handleMaxClick = () => {
-    setAmount(tokenBalance.toString());
+  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
```

Comment:
It either should not allow entering negative values or should validate on it and not allow clicking `Add Stake` then